### PR TITLE
Fix CoastlineFreeSlip boundary conditions for HydrostaticBoussinesqModel

### PR DESCRIPTION
**Pull request by blallen**
# Description

The current implementation didn't properly zero out the normal component of velocity. Updated these to follow equations 1.26 and 1.28 from the numerics section of the design docs. 

I have

- [x] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and run `julia .dev/format.jl`

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Ocean/HydrostaticBoussinesq/OceanBoundaryConditions.jl [hunk 1]
```diff
@@ -34,10 +33,39 @@ nothing needed to do since these are neumann BCs and no gradients here
...
 @inline function ocean_boundary_state!(::HBModel, ::CoastlineFreeSlip,
-                                       ::Union{Rusanov,
-                                               CentralNumericalFluxGradient},
+                                       ::Rusanov,
                                        Q⁺, A⁺, n⁻, Q⁻, A⁻, t)
-  return nothing
+    u⁻ = Q⁻.u
+    n = @SVector [n⁻[1], n⁻[2]]
+    
+    Q⁺.u = u⁻ - 2 * (n∘u⁻) * n
```
Comment (jkozdon) [inline src/Ocean/HydrostaticBoussinesq/OceanBoundaryConditions.jl:41]: Do you really want to use `∘` here for dot product? `∘` is a julia function for function composition.

Comment (blallen) [reply src/Ocean/HydrostaticBoussinesq/OceanBoundaryConditions.jl:41]: Ah, I was using it instead of · (\cdot) because it's easier to read. Didn't realize it already had a syntactical meaning. I can switch back to using \circ.

**Comment by blallen** [comment]
> This is what I would expect for zeroing out the normal velocity.
> 
> Can you put in a test?

The only real test we have right now would be running `test_divergence_free.jl` with FreeSlip instead of NoSlip boundary conditions. That's a good first check, but I'm not sure if it will give the level of testing needed. Any thoughts @jm-c ?

**Comment by jm-c** [comment]
The ShallowWater problem with either the Stommel or the Munk solution
(depending on FreeSlip/NoSlip BC on the side and on the bottom) should be 
a nice test, at least for Coastline BC.

**Comment by blallen** [comment]
> The ShallowWater problem with either the Stommel or the Munk solution (depending on FreeSlip/NoSlip BC on the side and on the bottom) should be a nice test, at least for Coastline BC.

Can we do this in 3D? That's the only way to test the BC for this model.

Also, I just pushed a bunch of preliminary work to enable easily swapping between FreeSlip and NoSlip

**Comment by jm-c** [comment]
@blallen  OK, I see your point regarding ShallowWater problem.

**Comment by blallen** [comment]
> Looks good to me. The Rusanov CoastlineFreeSlip BC is fixed for good (similar to sLinearDrag, in shallow_boundary_state).
> Regarding the CentralNumericalFluxGradient BC, it make sense to me but would be nice if @sandreza
> could also agree.
> And I re-run ocean_gyre.jl after switching to CoastlineFreeSlip and it looks reasonable.

For these tests, did you run with OceanFloorFreeSlip or OceanFloorNoSlip? I want to know so I can replicate the setup in the test files.

**Comment by jm-c** [comment]
I've just run keeping OceanFloorNoSlip.
This is different from what I run with MITgcm (to produce the little movie I sent the other day) 
which has FreeSlip @ bottom but noSlip on the coast.

**Comment by blallen** [comment]
> I've just run keeping OceanFloorNoSlip. This is different from what I run with MITgcm (to produce the little movie I sent the other day) which has FreeSlip @ bottom but noSlip on the coast.

Okay, I'm setting up the Azure tests to run just switching Coastline between FreeSlip and NoSlip, but will test all permutations of the boundary conditions on slurm CI.

File: experiments/OceanBoxGCM/ocean_gyre.jl [hunk 2]
```diff
@@ -88,4 +90,51 @@ function run_ocean_gyre(; imex::Bool = false)
...
+            CLIMA.HydrostaticBoussinesq.CoastlineFreeSlip(),
+            CLIMA.HydrostaticBoussinesq.OceanFloorFreeSlip(),
+            CLIMA.HydrostaticBoussinesq.OceanSurfaceNoStressForcing(),
+        ),
+    ]
+
+    for BC in boundary_conditions
+        run_ocean_gyre(imex = false, BC = BC)
+    end
+end
```
Comment (blallen) [inline experiments/OceanBoxGCM/ocean_gyre.jl:140]: @simonbyrne @kpamnany @charleskawczynski , do you guys have any ideas on how to easily switch between running all of these options for CI and only running a single setup interactively? 

Currently, this is the only place we can test that all the different boundary conditions are working... besides `test_ocean_gyre.jl` but I fear that running all of these on Azure would take too long and that the integration time is too short to really test the physics anyways

Comment (charleskawczynski) [reply experiments/OceanBoxGCM/ocean_gyre.jl:141]: Hm, I would suggest testing BCs with a coarse mesh and for a single timestep (assuming they're time-independent BCs)--if it's BCs that you're testing. At first glance, this looks like an entire simulation, in which case I would suggest breaking each case out into individual tests/experiments and running them in parallel on slurm. That said, we definitely need some sort of unit tests for BCs in general.

Comment (blallen) [reply experiments/OceanBoxGCM/ocean_gyre.jl:141]: But the boundary condition can run for a single time step successfully but not produce correct results if run for a longer length of time.  We need some way to catch these errors that are small per timestep but gradually accumulate.

**Comment by blallen** [comment]
bors try

File: experiments/OceanBoxGCM/ocean_gyre.jl (hunk 2 above)
Comment (charleskawczynski) [reply experiments/OceanBoxGCM/ocean_gyre.jl:141]: If the errors depend on the initial state, maybe we should supply an artificial initial state to check the BCs. Or apply a divergence-free random IC. So long the BCs are not time-dependent, a single step/BC application is all that should be needed to test the BCs.

Comment (kpamnany) [reply experiments/OceanBoxGCM/ocean_gyre.jl:141]: It'd be good to have two tests -- a basic and quick test that provides some confidence that things are working, and a detailed test that assures that things are working. The first will go in merge tests, the second in nightly (which is not up yet).

**Comment by kpamnany** [comment]
Do all the commits pass CI or can you squash?

**Comment by blallen** [comment]
Each commit should pass CI individually. 

aa4f07d7a327307f5417866ee5c7a11bbf15e856, e7109e68a00e735f8498a2e3625f383354e553dc, 439fb1ca0cf838932eca53234b9bfaafb2621c0b aren't functionally; they just rename things or move them around

5dcc80a5ce11b4688b3d8315159b3b029381e975, 2008614eb92b07eb9ee45d93443e9b874d5bde17, 43f0ade96d10b15459c2ca66dd8c2693c8602344 are all atomic changes towards the end goal:
- 5dcc80a5ce11b4688b3d8315159b3b029381e975 fixes the BC
- 2008614eb92b07eb9ee45d93443e9b874d5bde17 enables easily switching BCs but doesn't switch in tests
- 43f0ade96d10b15459c2ca66dd8c2693c8602344 actually tests multiple BCs in one test

**Comment by blallen** [comment]
Welp, didn't see that bors had started and decided to squash after all, here we go again

**Comment by charleskawczynski** [comment]
LGTM!

**Comment by blallen** [comment]
Bors r+
